Jmracek

j-mracek opened the pull request:
Applied proposed changes from last review

j-mracek committed:
Changed Dockerfile lauch-test to provide Apache plus reduction of steps.

Dockerfile do not build RPM and Libsolv git repos, only upgrade them from RawHide ropository.

This project is built on richdeps-docker available on https://github.com/shaded-enmity/richdeps-docker and step by step it will be modified for dnf-testing purpose of pull-requests.

j-mracek committed:
Added UPGRADE and repo creator, and renamed repo and feature

New test option upgrade for packages or all installed packages and test if packages where upgraded was added. Repository creator was added that use repo.ini that contains  information about Name, Version, Release, Architecture and other option like Requires in format that Rpmfluff use to created packages and repository.

To keep original test test-1.feature and test-2.feature was renamed to rich-1.feature and rich-2.feature . Repo test-1 was renamed to rich-1 .

j-mracek committed:
Original repository and features were renamed

To keep oroginal rich-dep repos and features they were renamed like test-1 to rich-1

j-mracek committed:
Added unupgraded staus for Then in test

j-mracek committed:
Added test-3 for test dnf upgrade with unspecified package

j-mracek committed:
In test_behave.py removed some PEP-8 violations

j-mracek committed:
Add dnf downgrade test

Added dnf downgrade test in test-4 and introduce downgrade action and downgraded state.

j-mracek committed:
Enlarged repo/test-1 and test-1 with more test steps

Test-1 and repo/test-1 test more cases with install and remove were various Requires play role including install of package that cannot meet Requires.

j-mracek committed:
Changed description of test-1

j-mracek committed:
Enlarged test-1 and repo/test-1

Test-1 additionally controls that requires package is not removed if other installed package need it.

j-mracek committed:
Added present state and changed absent state

Present state indicate that package was not installed or removed and it is present in RPM db. Absent state now also control if package is absent in RPM db.

j-mracek committed:
Added dnf db check for states install, removed, absent, present

j-mracek committed:
Enhanced repo/upgrade_1 and test-1, test-2

Redo/upgrade_1 and test-2 test more cases during upgrade of packages. In test-1 added step to install installed package.

j-mracek committed:
Modified test-2. Two packages upgraded with dependencies in one step.

j-mracek committed:
Test-2, test-3, and repo/upgrade_1 enhanced for test dnf upgrade.

Added step and rpm where TestK cannot be upgraded due requires TestM-1.0.0-3

j-mracek committed:
Changed description in test-2 and test-3, and added one step in test-2

j-mracek committed:
Changed relative path detection, new dir shared

In test-launcher.py relative path detection was changed to avoid problem if run path is not dir with test-launcher.py. New dir is shared with docker image.

Added new initial file for docker image setup

j-mracek committed:
Added initials for Jenkins test

Added import, and parser of ci-dnf-stack.log file. Added dnf-nightly upgrade. Added upgrade-to dnf-tested version.

j-mracek committed:
Enhanced dnf package parser and added new temp dir dnf.repo

j-mracek committed:
Fixed multiple packages with same version in repository in initial.py

If more packages of DNF with same version in repository, it find the correct one by git number.

j-mracek committed:
Moved *.repo to temp dir and behave runs only if initial.py runs OK

Repos are moved from mv /etc/yum.repos.d/* to /temp/dnf.repo/ and behave runs only if initial.py runs with exit code 0. All changes were made in launch-test file.

j-mracek committed:
Renamed tests

To enable reintroduction original tests from Pavel Odvody and allow more tests, I renamed tests.

Test1 to test6 were renamed as test-A to Test-F and repository test-1 was renamed as test-5

j-mracek committed:
Reworked log file parser

Parser were changed that test-laucher.py provides a key information about DNF version for upgrade from log for each test.

Parionaly, parser were moved from initial_settings/initial.py to test-laucher.py and DNF API is used insted of re.

j-mracek committed:
Added new WHEN def for commands to cli and new test-G

The new WHEN statement perform direct command line commands. This option allow high variability of options for tests.

Test-G was introduced with new WHEN statement.

j-mracek committed:
Removed colored output

Colored output defined in test-launcher.py was removed due to incompatibility with logging module.

j-mracek committed:
Minor changes in test-launcher.py to follow PEP-8 formating

jsilhan commented on features/steps/test_behave.py:
```diff
@@ -1,108 +1,229 @@
+    """ Gets all installed packages in the system with version to check that dnf has same data about installed packages"""
+    pkgverstr = subprocess.check_output(['dnf', 'repoquery', '--installed', '-C', '--queryformat', '%{name}.%{version}.%{release}\n'])
+    pkgverstr = pkgverstr.splitlines()
+    return pkgverstr[1:]
```
this looks hacky (probably because of metadata information), add `-q` to the command instead, please

jsilhan commented on features/steps/test_behave.py:
```diff
@@ -1,108 +1,229 @@
+    return subprocess.check_call(['dnf'] + flags + cmd, stdout=subprocess.PIPE)
+
+
+def execute_dnf_command_notinstall(cmd, reponame):
```
I had objection to `execute_dnf_command` and `execute_dnf_command_notinstall` functions. Reuse this if needed https://github.com/shaded-enmity/richdeps-docker/commit/9a70fd3edfe06ea71e908e3bd62193fef1c25e63

jsilhan commented on features/steps/test_behave.py:
```diff
@@ -1,108 +1,229 @@
+    assert repo
+    context.repo = repo
+    assert os.path.exists('/var/www/html/repo/' + repo)
+    subprocess.check_call(['rm -rf /repo/*'], shell=True)
```
why have you replace Python code (`a = [os.remove(p) for p in os.listdir('/repo')]`) with system call?

jsilhan commented on features/steps/test_behave.py:
```diff
@@ -1,108 +1,229 @@
+            else:
+                execute_dnf_command([action] + split(pkg), context.repo)
+        elif action == 'notinstall':
+            exit_code = execute_dnf_command_notinstall(["install"] + split(pkg), context.repo)
```
No, see above ^

jsilhan commented on features/steps/test_behave.py:
```diff
@@ -1,108 +1,229 @@
+            assert exit_code != 0
+        elif action == 'autoremove':
+            subprocess.check_call(['dnf', '-y', action], stdout=subprocess.PIPE)
+        else:
```
replace with `elif`, please

jsilhan commented on features/steps/test_behave.py:
```diff
@@ -1,108 +1,229 @@
+        elif action == 'autoremove':
+            subprocess.check_call(['dnf', '-y', action], stdout=subprocess.PIPE)
+        else:
+            execute_dnf_command([action] + split(pkg), context.repo)
```
``` python
else:
    raise AssertError(..)
```

jsilhan commented on features/steps/test_behave.py:
```diff
@@ -1,108 +1,229 @@
-  assert pkg
-  context.pre_packages = get_package_list()
-  assert context.pre_packages
+    assert action in ["install", "remove", "upgrade", "downgrade", "notinstall", "autoremove", "upgrade-to"]
```
delete in favor of https://github.com/jsilhan/richdeps-docker/pull/2/files#r44656331

jsilhan commented on features/steps/test_behave.py:
```diff
@@ -1,108 +1,229 @@
+    return [find_pkg(p) for p in pkgs]
+
 
 def get_package_list():
```
rename to `get_rpm_package_list`

jsilhan replied:
this smells to me to be redundant to `get_rpm_package_version_list`

jsilhan commented on features/steps/test_behave.py:
```diff
@@ -1,108 +1,229 @@
+    return pkgstr.splitlines()
+
+
+def get_package_version_list():
```
rename to `get_rpm_package_version_list`

jsilhan commented on features/steps/test_behave.py:
```diff
@@ -1,108 +1,229 @@
+    return pkgverstr.splitlines()
+
+
+def get_package_dnf_version_list():
```
rename to `get_dnf_package_version_list`

jsilhan commented on features/test-A.feature:
```diff
@@ -0,0 +1,30 @@
+ And package "TestL, TestM" should be "present"
+ And package "TestL, TestM" should be "unupgraded"
+ When I "remove" a package "TestL" with "dnf"
+ Then package "TestL, TestM" should be "removed"
```
add new empty line, please

jsilhan replied:
Can these tests `when` statements be separated by new line (if the state is clean and does not depend on previous statements)

jsilhan commented on repo_create/spec_creater.py:
```diff
@@ -0,0 +1,62 @@
+        os.makedirs(f)
+
+
+def rpm_init(function, args):
```
?

jsilhan commented on repo_create/test-1/spec.txt:
```diff
@@ -0,0 +1,14 @@
+TestA	1	TestB
```
This shoudln't be there, right?

jsilhan commented on repo_create/test-1/spec/TestA-1.spec:
```diff
@@ -0,0 +1,26 @@
+Release:	1
+Summary:	Richdeps test package
+
+#Group:
```
remove this, please

jsilhan commented on repo_create/test-1/spec/TestA-1.spec:
```diff
@@ -0,0 +1,26 @@
+Source0:	testdata.tar.gz
+BuildArch:  noarch
+
+#BuildRequires:
```
remove this, please

jsilhan commented on repo_create/upgrade_1/spec.txt:
```diff
@@ -0,0 +1,28 @@
+TestA	1	TestB
```
erase this file

jsilhan commented on test-launcher.py:
```diff
@@ -1,55 +1,53 @@
 
-def color_text(c, text):
-  return "{}{}{}".format(c, text, Colors.ENDC)
+def log_parser():
```
isn't `get_dnf_testing_version()` better name for that?

jsilhan commented on test-launcher.py:
```diff
@@ -1,55 +1,53 @@
-    print(red_text('-'*80))
-    print(header_text("Container returned non zero value({})".format(rc)))
-    print(red_text('-'*80))
+if len(sys.argv) == 1:
```
this and all below shoudl be inside `if __name__ != "__main__":` block

jsilhan commented on test-launcher.py:
```diff
@@ -1,55 +1,53 @@
 if not r:
-  print(green_text('OK'))
+    print('OK')
+else:
```
print("FAIL")

jsilhan commented on repo_create/spec_creater.py:
```diff
@@ -0,0 +1,62 @@
+
+
+spec_rpm_creator()
+create_repo()
```
I advise you to use rpmfluff to avoid doing your own parser but you still did that again :). Attach how the test definition look like, please.

j-mracek committed:
When and than handle exit codes, reworked cleaning /repo and minor changes

Updated statement 'When I execute command {command} with {result}' now handle results 'success' and 'fail'. The When statement now replace notinstall function including description in test-A.feature.

New Then statement handle comparison of command exit code and expected value.

Reworked removing all files and folders from /repo

Minor changes in assertion of allowed actions and statement

Minor changes in description of tests where new line was inserted to separate functional blocks in tests

j-mracek committed:
Use Fedora:23 to build docker images plus minor changes

Fedora:23 images replaced Fedora:rawhide to decrease problems during build

Use Deltarpm were disable to fix problem during docker image build

Minor changes – all install and upgrade steps were accumulated

j-mracek committed:
Replaced repo test-1 and upgrade_1 with repo created by rpmfluff

j-mracek committed:
Minor –  renamed variable pkg and introduction if __name__ ==…

Variable pkg represented pkgs separated by comma therefore it was renamed to pkgs in test_behave.py .

Statement if __name__ == "__main__" was introduced in test-launcher.py

jsilhan commented on README.md:
```diff
@@ -1,2 +1,60 @@
+
+Possible package managers: dnf, rpm (pkcon soonish)
+
+Possible states: installed, removed, absent
```
:thumbsup:

jsilhan commented on features/steps/test_behave.py:
```diff
@@ -0,0 +1,244 @@
+
+def package_absence(pkg, list_ver):
+    """ Select package versions """
+    found_pkgs = [x for x in list_ver if re.search('^' + pkg, x)]
```
`... if x.startswith(pkg)...` (you know for the next time, regexps are slower than string comparison)

jsilhan commented on features/steps/test_behave.py:
```diff
@@ -0,0 +1,244 @@
+    if not isinstance(pkg, list):
+        pkg = [pkg]
+    if action == "remove":
+        action = RPM_ERASE_FLAGS
```
this should be new variable. E.g. `rpm_command` - it's more error prone when you change meaning of the variable. let them be immutable if possible

jsilhan commented on features/steps/test_behave.py:
```diff
@@ -0,0 +1,244 @@
+    return sorted(a) == sorted(b)
+
+
+def split(pkgs):
```
is this really needed?

jsilhan commented on initial_settings/initial.py:
```diff
@@ -0,0 +1,35 @@
+        return subprocess.check_call(['dnf', 'upgrade', '-y', '--disablerepo=*', '--enablerepo=dnf-nightly'])
+
+    def upgrade(self, pkg):
+        return subprocess.check_call(['dnf', 'upgrade-to', '-y', '--disablerepo=*',
```
what about

``` python
"dnf upgrade -y --disablerepo=* --repofrompath=dnf-pull-requests,https://copr-be.cloud.fedoraproject.org/results/rpmsoftwaremanagement/dnf-pull-requests/fedora-rawhide-x86_64/ --enablerepo=dnf-pull-requests " + pkg
```

j-mracek replied:
Unfortunately repo dnf-pull-requests is needed for dnf_version() to search for exact name of dnf package. Introduction of repo in upgrade step is too late

jsilhan commented on initial_settings/initial.py:
```diff
@@ -0,0 +1,35 @@
+import dnf
+
+
+class DnfUpgrageTo():
```
delete, see lower why

jsilhan replied:
rename to `DnfEnvSetup`, please

jsilhan commented on initial_settings/initial.py:
```diff
@@ -0,0 +1,35 @@
+            assert len(query) == 1
+            return str(query[0])
+
+    def upgrade_nightly(self):
```
upgrade_dnf_dependencies_from_nightly

jsilhan commented on initial_settings/initial.py:
```diff
@@ -0,0 +1,35 @@
+                    'rpmsoftwaremanagement/dnf-nightly/fedora-rawhide-x86_64/\nenabled=1\ngpgcheck=0')
+        return subprocess.check_call(['dnf', 'upgrade', '-y', '--disablerepo=*', '--enablerepo=dnf-nightly'])
+
+    def upgrade(self, pkg):
```
upgrade_dnf_from_pull_request

jsilhan commented on initial_settings/initial.py:
```diff
@@ -0,0 +1,35 @@
+
+
+class DnfUpgrageTo():
+    def create_repo(self):
```
should not be needed, see https://github.com/jsilhan/richdeps-docker/pull/2/files#r45382081

j-mracek replied:
Unfortunately it is needed for dnf_version() function. It search for exact name of dnf package for upgrade-to command.

jsilhan commented on repo_create/test-1.py:
```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/python -tt
```
#!/usr/bin/python3 -tt

jsilhan replied:
I am not sure what is target Python versoin of whole project but we should use Python 3

jsilhan commented on repo_create/test-1.py:
```diff
@@ -0,0 +1,96 @@
+
+work_file = os.path.realpath(__file__)
+work_dir = os.path.dirname(work_file)
+file_base_mane = work_file.rsplit('/', 1)[1].rsplit('.', 1)[0]
```
if we use Python3, then take advantage of [pathlib](https://docs.python.org/3/library/pathlib.html)

jsilhan commented on repo_create/test-1.py:
```diff
@@ -0,0 +1,96 @@
+
+shutil.rmtree(temp_dir)
+
+print("DNF repo is located at %s" % repo.repoDir)
```
this is whole repo definition for all the tests?

j-mracek committed:
Minor: re was replaced startwith and renamed variable

To increase speed of script re was replaced with  x.startwith() in test_behave.py .

Variable action was renamed to rpm_command in test_behave.py .

j-mracek committed:
Updated README.md
